### PR TITLE
chore(RHINENG-2211): React 18 & Patternfly 5 Upgrades

### PR DESCRIPTION
**Pull request by LiorKGOW**
# Description

This PR upgrades this application to React 18 & Patternfly 5 (along with some necessary packages)

Associated Jira tickets: 
[#RHINENG-2211](https://issues.redhat.com/browse/RHINENG-2211)
[#RHINENG-5939](https://issues.redhat.com/browse/RHINENG-5939)

As mentioned in [this slack thread](https://redhat-internal.slack.com/archives/C02T90PLJ5V/p1705484895906979), we agreed on not upgrading `redhat-cloud-services/frontend-components-config`, 
therefore I did not change it in this PR

# How to test the PR

Have a look that the application works as expected

# Checklist:

- [x] The commit message has the Jira ticket linked
- [x] PR has a short description
- [x] Upgrade to React 18
- [x] Upgrade to Patternfly 5
- [X] Fix failing tests
- [x] Fix commits

### CODE REVIEW

**Comment by jmbuscemi** [comment]
I would just like to request that Teamnado's QE team gets a chance to run our Activation Key regression tests once this makes it's way to Stage and before it goes to Prod so that we can assure nothing there is broken.

You can ping myself or Luke in our Slack channel and we'll get the QE team to run the tests.

**Comment by LiorKGOW** [comment]
> I would just like to request that Teamnado's QE team gets a chance to run our Activation Key regression tests once this makes it's way to Stage and before it goes to Prod so that we can assure nothing there is broken.
> 
> You can ping myself or Luke in our Slack channel and we'll get the QE team to run the tests.

I will ping you once this PR gets merged @jmbuscemi 👍🏼

**Comment by LiorKGOW** [comment]
This PR is ready for review ✌🏼

File: src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap [hunk 1]
```diff
@@ -3,54 +3,70 @@
...
+        </div>
+      </div>
       <div
-        class="pf-c-empty-state__body"
+        class="pf-v5-c-empty-state__body"
       >
         Contact your organization administrator(s) for more information or visit 
         <a
-          href="./settings/my-user-access"
+          href="./iam/my-user-access"
```
Comment (bastilian) [inline src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: Is this change intended?

Comment (bastilian) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: hm. this is a snapshot. I'm wondering where this comes from and if this is also in the actually rendered UI. And if it is actually correct.

Comment (LiorKGOW) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: let me have a look

Comment (LiorKGOW) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: @bastilian This change came from this commit:
https://github.com/RedHatInsights/sed-frontend/pull/272/commits/1d4fe7cdbae4c010fa53ffc1f02cf6a58e6ded7a 
But I did not make any changes to `Authentication.js` here so I guess it came from a different commit and I just updated the snapshot here

I checked previous commits, and I only made relevant changes in those commits:
1. This is the only commit where changes were made to `Authentication.js` & `Authentication.test.js` files, but those changes came after the snapshot update you mentioned, so they couldn't have came up from this commit: https://github.com/RedHatInsights/sed-frontend/pull/272/commits/53a5bf9dbfe01af4f72327af0ccd5ddd6c455d4f#diff-a194d08568455b4d77eb5df13c98260410d3ed5375a39aee43331d657dd85489 
2. Here was the first snapshot update to the snapshot file you mentioned, but I did not make any previous changes to `Authentication.js` & `Authentication.test.js` files, so it must have been caused from a previous package update, but I am not sure which one caused this change.
https://github.com/RedHatInsights/sed-frontend/pull/272/commits/f01bac6dd852a4b70b68d4b23200ea39f1349426#diff-f6381abb627b767d59d1844907725850526ea230abf57154c7492a365c0e5a20

Maybe this change came from the `config/setupTests.js` change in the same commit ? or a previous package update ?

Comment (LiorKGOW) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: I found it in the UI

This is the page it is found on:
image

And this is the link it sends you to when you click [My User Access](https://stage.foo.redhat.com:1337/iam/my-user-access):
image

It seems to me that this is the functionality we want when we click this link @bastilian

File: .cache_iifckc
Comment (mkholjuraev) [inline .cache_iifckc:1]: I believe we can remove this cache file generated by pf-codemods as it will not be used anymore

File: src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap (hunk 1 above)
Comment (jmbuscemi) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: That first screenshot is what we need to show in Activation Keys when a user doesn't have access.

If you need help with any of the Activation Key pieces let us know and we can take a look.

File: .cache_iifckc
Comment (LiorKGOW) [reply .cache_iifckc:1]: I removed it 👍🏼 
Thanks @mkholjuraev

File: src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap (hunk 1 above)
Comment (LiorKGOW) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: @jmbuscemi Does the link work as expected ? 
We were just wondering if the code works as it should (because the link in the snapshot changed)

Comment (jmbuscemi) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: @jlgre when you get some time can you look into Activation Keys w.r.t. this PR?

Comment (lukegrn) [reply src/Components/Authentication/__tests__/__snapshots__/Authentication.test.js.snap:46]: I believe that is the correct link  - I don't even think that's a component that we control.

settings/my-user-access breaks anyway as far as I can see, so the iam one seems right

**Comment by lukegrn** [comment]
Haven't reviewed the code yet, but upon running locally I see an error in the activation keys create wizard. When the 4th step is entered and something has been selected for SLA on the 3rd step, the modal crashes upon step advance.

**Comment by gkarat** [comment]
@LiorKGOW, I tried to reproduce the bug that @jlgre found and I see that there is a non-string value we set for `sla` through `setSla` function: https://github.com/LiorKGOW/sed-frontend/blob/react-18-upgrade/src/Components/Pages/SetSystemPurposePage.js#L73. I checked the new interface for `onChange` for FormSelect and there is indeed the first argument is of event type. So this would need to be rewritten to smth like:
 ```
            <FormSelect onChange={(event, value) => setSla(value)} value={sla} id="activation-key-sla">
```

**Comment by gkarat** [comment]
The warnings in the console can be helpful and indicate if there are some props of a wrong type passed to components/functions.

**Comment by LiorKGOW** [comment]
Thank you for having a look @gkarat @jlgre 
I have addressed the issue you mentioned @jlgre and added the changes you suggested @gkarat in this commit: [(fix): Change onChange arguments of FormSelect](https://github.com/RedHatInsights/sed-frontend/pull/272/commits/472f1d18883adc4822adbe4aae864950973f66ec)

Also, I fixed two console errors in the last commit

The PR is ready for another review 😄

**Comment by LiorKGOW** [comment]
BTW, I thought we can also do this in the following way:
Define a more generic function that gets the necessary setter function and returns a fitting onChange function and provide it with the necessary setter function for each FormSelect
```
  const createOnChangeFunc = (setterFunc) => {
    return (_event, value) => setterFunc(value);
  };
  .
  .
  .
  <FormSelect
              onChange={createOnChangeFunc(setRole)}
              value={role}
              id="activation-key-role"
            >
```
What do you think @gkarat ?

File: src/Components/ActivationKey/SystemPurposeCard.js [hunk 2]
```diff
@@ -41,7 +40,21 @@ const SystemPurposeCard = (props) => {
...
+            <>
+              {user.rbacPermissions.canWriteActivationKeys ? (
+                <ButtonWrapper />
+              ) : (
+                <NoAccessPopover content={ButtonWrapper} />
+              )}
+            </>
+          ),
+          hasNoOffset: false,
+          className: undefined,
```
Comment (mkholjuraev) [inline src/Components/ActivationKey/SystemPurposeCard.js]: Any reason to have className: undefined?

File: src/Components/ActivationKey/WorkloadCard.js [hunk 3]
```diff
@@ -32,7 +31,17 @@ const WorkloadCard = (props) => {
...
-      <CardHeader>
+      <CardHeader
+        actions={{
+          actions: (
+            <>
+              <EditButton onClick={actionHandler} />
+            </>
+          ),
+          hasNoOffset: false,
+          className: undefined,
```
Comment (mkholjuraev) [inline src/Components/ActivationKey/WorkloadCard.js]: here as well

**Review by mkholjuraev — COMMENTED** [review]
there is this one that needs a fix before we can merge the PR. When I try to add a new repo to my activation keys and try to filter by name from the dropdown, the modal gets closed unexpectedly or the page gets broken. Ref: https://redhat-internal.slack.com/archives/C02T90PLJ5V/p1708335619534349?thread_ts=1708267864.925449&cid=C02T90PLJ5V

**Comment by LiorKGOW** [comment]
Thank you for the review @mkholjuraev ! 

I have fixed the issue you mentioned when trying to filter the repositories by name, it was just a small change in the onChange function. In PF5 SearchInput, when the onChange function is called, the first argument is the event and the second argument is the changed value. 
It should work now

Also, I changed the classNames to be more meaningful
My guess is these classNames came from the 'pf-codemod' changes

The PR is ready for another review :)

File: src/Components/ActivationKey/SystemPurposeCard.js (hunk 2 above)
Comment (LiorKGOW) [reply src/Components/ActivationKey/SystemPurposeCard.js]: No haha 
My guess is these classNames came from the 'pf-codemod' changes

I changed the classNames to be more meaningful

**Review by mkholjuraev — APPROVED** [review]
LGTM! I had another quick look and did not encounter issues. @LiorKGOW Thank you! 

Tests need a fix though

**Comment by LiorKGOW** [comment]
I removed Cypress from the pipeline because there are no Cypress tests anymore (I converted the last one to RTL)
The test is passing now @mkholjuraev 🥳

**Comment by LiorKGOW** [comment]
@jmbuscemi The PR is merged
You can now find this on stage
